bundle/regal/ast.rego

```diff
@@ -125,17 +137,61 @@ _find_vars(path, value, last) := _find_every_vars(path, value) if {
 	value.domain
 }
 
+_find_vars(_, value, _) := _find_set_or_array_comprehension_vars(value) if {
+	value.type in {"setcomprehension", "arraycomprehension"}
+}
+
+_find_vars(path, value, _) := _find_object_comprehension_vars(value) if {
+	value.type == "objectcomprehension"
+}
+
 # METADATA
 # description: |
 #   traverses all nodes under provided node (using `walk`), and returns an array with
-#   all variables declared via assignment (:=), `some` or `every`
+#   all variables declared via assignment (:=), `some`, `every` and in comprehensions
 find_vars(node) := [var |
 	# regal ignore:function-arg-return
 	walk(node, [path, value])
 
 	some var in _find_vars(path, value, regal.last(path))
 ]
 
+_function_arg_names(rule) := {arg.value |
+	some arg in rule.head.args
+	arg.type == "var"
+}
+
+# METADATA
+# description: |
+#  finds all vars declared in `rule` *before* the `location` provided
+#  note: this isn't 100% accurate, as it doesn't take into account `=`
+#  assignments / unification, but it's likely good enough since other rules
+#  recommend against those
+find_vars_in_local_scope(rule, location) := [var |
+	some var in find_vars(rule)
+	_before_location(var, location)
```

Comment:
Are we working with `ast.Module`s that have passed through the compiler...? Because IIRC we do some reordering there based on output variables, that might be relevant to consider. I suppose it could be possible that some policies are valid for OPA but icky in Regal...? 🤔

Reply:
No, we don't leverage the compiler in Regal at this point, but it's been considered for having something like the strict mode checks included. At that point we'd probably need to revisit some assumptions like you're on to 😅